frontend/canonical.php

```diff
@@ -0,0 +1,261 @@
+		if ( 1 == $this->options['force_lang'] && is_attachment() && isset( $_GET['attachment_id'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+			return;
+		}
+
+		/*
+		 * If the default language code is not hidden and the static front page url contains the page name,
+		 * the customizer lands here and the code below would redirect to the list of posts.
+		 */
+		if ( is_customize_preview() ) {
+			return;
+		}
+
+		if ( empty( $requested_url ) ) {
+			$requested_url = pll_get_requested_url();
+		}
+
+		if ( ( is_single() || is_page() ) && ! is_front_page() ) {
+			$post = get_post();
+			if ( $post instanceof WP_Post && $this->model->is_translated_post_type( $post->post_type ) ) {
+				$language = $this->model->post->get_language( (int) $post->ID );
+			}
+		}
+
+		if ( ! empty( $wp_query->tax_query ) ) {
+			if ( $this->model->is_translated_taxonomy( $this->get_queried_taxonomy( $wp_query->tax_query ) ) ) {
+				$term_id = $this->get_queried_term_id( $wp_query->tax_query );
+				if ( $term_id ) {
+					$language = $this->model->term->get_language( $term_id );
+				}
+			}
+		}
+
+		if ( $wp_query->is_posts_page ) {
+			$page_id = get_query_var( 'page_id' );
+			if ( ! $page_id ) {
+				$page_id = get_queried_object_id();
+			}
+			if ( $page_id && is_numeric( $page_id ) ) {
+				$language = $this->model->post->get_language( (int) $page_id );
+			}
+		}
+
+		if ( 3 === $this->options['force_lang'] ) {
+			$requested_host = wp_parse_url( $requested_url, PHP_URL_HOST );
+			foreach ( $this->options['domains'] as $lang => $domain ) {
+				$host = wp_parse_url( $domain, PHP_URL_HOST );
+				if ( $requested_host && $host && ltrim( $requested_host, 'w.' ) === ltrim( $host, 'w.' ) ) {
+					$language = $this->model->get_language( $lang );
+				}
+			}
+		}
+
+		if ( empty( $language ) ) {
+			$language = $this->curlang;
+			$redirect_url = $requested_url;
+		} else {
+			$redirect_url = $this->redirect_canonical( $requested_url, $language );
+			$redirect_url = $this->options['force_lang'] ?
+				$this->links_model->switch_language_in_link( $redirect_url, $language ) :
+				$this->links_model->remove_language_from_link( $redirect_url ); // Works only for default permalinks.
+		}
+
+
+		/**
+		 * Filters the canonical url detected by Polylang.
+		 *
+		 * @since 1.6
+		 *
+		 * @param string|false $redirect_url False or the url to redirect to.
+		 * @param PLL_Language $language The language detected.
+		 */
+		$redirect_url = apply_filters( 'pll_check_canonical_url', $redirect_url, $language );
+
+		if ( ! $redirect_url || $requested_url === $redirect_url ) {
+			return $requested_url;
+		}
+
+		if ( ! $do_redirect ) {
+			return $redirect_url;
+		}
+
+		// Protect against chained redirects.
+		if ( $redirect_url === $this->check_canonical_url( $redirect_url, false ) && wp_validate_redirect( $redirect_url ) ) {
+			wp_safe_redirect( $redirect_url, 301, POLYLANG );
```

Comment:
Do you think we could perform the redirection in an other method, this way we could easily test `check_canonical_url()` with `$do_redirect` set to `true` and use a mock or something to avoid  the redirection in unit test. I remember we had some issue with this part of the code with infinite redirection.

Reply:
I took care to position the [problematic condition](https://github.com/polylang/polylang/pull/1074/files#diff-1f56dc45bed29b524bb07f3e0f6588c99486c88cf33178dd16ec12f984b18521R147) before the `$do_redirect` so we should not be trapped the same way.

Using a dedicated method for redirect was discussed in the past but (temporarily?) abandonned. See https://github.com/polylang/polylang-pro/pull/460. If we restart the idea, we should do it everywhere at once.